File: src/sentinel/strategy/sma.py
"""SMA-crossover reference strategy — target-position form.

Fast SMA above slow -> desired stance LONG. Fast below -> FLAT. That's the
whole rule: the strategy states what it wants to hold given the data, every
bar. It is NOT edge-based — it reports the current desired stance, and the
runner decides whether any order is needed. Warming up -> no opinion (None),
so the runner leaves any existing position untouched until the strategy has
enough data.

Deliberately simple: the point of Sentinel is execution integrity, not alpha.
"""

from __future__ import annotations

from collections import deque
from decimal import Decimal

from .base import Decision, Stance


class SmaCross:
    def __init__(self, fast: int = 5, slow: int = 20, short: bool = False,
                 stop_floor_pct: Decimal = Decimal("0.005")) -> None:
        if fast >= slow:
            raise ValueError("fast period must be shorter than slow")
        # short=True -> stop-and-reverse: SHORT below the cross instead of FLAT,
        # so the strategy is always in the market and can profit in a bear leg
        # (needs a venue that allows shorting; spot clamps SHORT to FLAT).
        self._short = short
        # The strategy's protective stop is the SLOW SMA (the trend line): a long
        # is wrong once price falls back to it. Floored so a near-cross entry
        # (price hugging the SMA) doesn't hand the risk layer a ~0 stop and size
        # itself into the leverage cap.
        self._stop_floor = stop_floor_pct
        self.name = f"sma-{'ls' if short else 'cross'}({fast}/{slow})"
        self.fast_period = fast     # public: the chart overlays these windows
        self.slow_period = slow
        self._fast_n = fast
        self._slow_n = slow
        self._prices: deque[Decimal] = deque(maxlen=slow)

    def view_spec(self) -> dict:
        """How the terminal should render this strategy: which detail values
        matter in the panel, and what to draw on the chart. The UI is generic;
        the strategy owns its own presentation."""
        return {
            "rows": [
                {"label": "Fast SMA", "key": "fast", "kind": "price"},
                {"label": "Slow SMA", "key": "slow", "kind": "price"},
            ],
            "overlays": [
                {"kind": "line", "key": "fast",
                 "label": f"SMA {self.fast_period}", "color": "#fbbf24"},
                {"kind": "line", "key": "slow",
                 "label": f"SMA {self.slow_period}", "color": "#a78bfa"},
            ],
        }

    def reset(self) -> None:
        """Drop all accumulated bars — the indicator must forget when the
        underlying timeframe changes, else the SMA blends two timeframes'
        closes into a meaningless average."""
        self._prices.clear()

    def on_bar(self, close: Decimal) -> Decision:
        self._prices.append(close)
        if len(self._prices) < self._slow_n:
            return Decision(None, {"warming_up": True,
                                   "have": len(self._prices),
                                   "need": self._slow_n})

        prices = list(self._prices)
        fast = sum(prices[-self._fast_n:]) / self._fast_n
        slow = sum(prices) / self._slow_n
        below = Stance.SHORT if self._short else Stance.FLAT
        stance = Stance.LONG if fast > slow else below
        detail = {"fast": str(round(fast, 2)), "slow": str(round(slow, 2))}
        if stance in (Stance.LONG, Stance.SHORT):
            # Stop geometry: distance from price back to the slow SMA (the trend
            # line), floored at stop_floor_pct of price. The risk layer sizes and
            # enforces the SL/TP off this.
            gap = abs(close - slow)
            detail["stop_dist"] = str(max(gap, close * self._stop_floor))
        return Decision(stance, detail)

File: go/internal/strategy/sma.go
// Package strategy is the part the OMS deliberately does not contain.
//
// A strategy declares a desired STANCE each closed bar; the runner reconciles
// the actual position to it. That target-position model — rather than acting on
// signal edges — is what makes it crash-safe: on restart you bring actual into
// line with desired and a pre-existing position is accounted for by
// construction, with no memory of which edge created it.
//
// Strategies here are pure and deterministic: a bar in, a stance out, no I/O,
// no clock, no randomness. That is what makes them unit-testable and replayable
// — feed the same bars, get the same stances. Orders reach the venue through
// the same control socket a human uses, so a strategy faces exactly the guards
// a human does. A bad strategy trades badly; it cannot corrupt state.
package strategy

import "github.com/gauryvg98/sentinel-oms/go/internal/fixed"

// Stance is the exposure a strategy wants to hold.
type Stance int

const (
	// NoOpinion is warming up: the runner does nothing at all, so a warm-up
	// never flattens a position that is already open.
	NoOpinion Stance = iota
	// Flat wants no exposure.
	Flat
	// Long wants positive exposure.
	Long
	// Short wants negative exposure. Perps only; a spot venue clamps it flat.
	Short
)

func (s Stance) String() string {
	switch s {
	case Long:
		return "LONG"
	case Short:
		return "SHORT"
	case Flat:
		return "FLAT"
	default:
		return "NONE"
	}
}

// Decision is one bar's answer.
type Decision struct {
	Stance Stance
	// Fast and Slow are the two averages, for the operator and the chart.
	Fast, Slow fixed.Dec
	// StopDist is how far price is from the trend line, floored and capped.
	// Zero when there is no directional stance to protect.
	StopDist fixed.Dec
	// Have and Need are non-zero only while warming up.
	Have, Need int
}

// SmaCross is the reference strategy: fast average above slow means hold long,
// below means flat — or short, if this is a venue that allows it.
//
// The point of this system is execution integrity, not alpha. The rule is
// deliberately the simplest thing that still has to be executed correctly.
type SmaCross struct {
	fast, slow int
	short      bool
	stopFloor  fixed.Dec
	stopCap    fixed.Dec
	prices     []fixed.Dec
}

// Config is how an SmaCross is built.
type Config struct {
	Fast, Slow int
	// Short turns the rule into stop-and-reverse: short below the cross rather
	// than flat, so it is always in the market.
	Short bool
	// StopFloorPct keeps a near-cross entry from handing the risk layer a stop
	// of almost nothing, which would size itself straight into the leverage cap.
	StopFloorPct fixed.Dec
	// StopCapPct keeps a trend that has run far from its average from producing
	// a 3-5% stop. Trend-following cuts losses short. Zero disables the cap and
	// uses the raw distance to the slow average.
	StopCapPct fixed.Dec
}

// New builds a strategy, or reports why the configuration is not one.
func New(c Config) (*SmaCross, error) {
	if c.Fast < 1 {
		return nil, ErrConfig{"fast period must be at least one bar"}
	}
	if c.Fast >= c.Slow {
		return nil, ErrConfig{"fast period must be shorter than slow"}
	}
	return &SmaCross{
		fast:      c.Fast,
		slow:      c.Slow,
		short:     c.Short,
		stopFloor: c.StopFloorPct,
		stopCap:   c.StopCapPct,
		prices:    make([]fixed.Dec, 0, c.Slow),
	}, nil
}

// ErrConfig is a strategy that could never have run.
type ErrConfig struct{ Why string }

func (e ErrConfig) Error() string { return "strategy: " + e.Why }

// Name identifies the strategy in a log.
func (s *SmaCross) Name() string {
	if s.short {
		return "sma-ls"
	}
	return "sma-cross"
}

// Reset drops every accumulated bar.
//
// The indicator has to forget when the timeframe changes, or the average blends
// two timeframes' closes into a number that describes neither.
func (s *SmaCross) Reset() { s.prices = s.prices[:0] }

// OnBar takes one closed bar and returns the stance it now wants.
func (s *SmaCross) OnBar(close fixed.Dec) Decision {
	s.prices = append(s.prices, close)
	if len(s.prices) > s.slow {
		s.prices = s.prices[len(s.prices)-s.slow:]
	}
	if len(s.prices) < s.slow {
		return Decision{Stance: NoOpinion, Have: len(s.prices), Need: s.slow}
	}

	fast := mean(s.prices[len(s.prices)-s.fast:])
	slow := mean(s.prices)

	stance := Flat
	if s.short {
		stance = Short
	}
	if fast.Raw() > slow.Raw() {
		stance = Long
	}

	d := Decision{Stance: stance, Fast: fast, Slow: slow}
	if stance == Long || stance == Short {
		// Anchored on the slow average — the trend line a long is wrong once
		// price falls back to — then floored and capped into a tight band, so
		// size and stop stay consistent however far the trend has run.
		gap := close.Sub(slow).Abs()
		if floor := close.Mul(s.stopFloor); floor.Raw() > gap.Raw() {
			gap = floor
		}
		if s.stopCap.Raw() > 0 {
			if cap := close.Mul(s.stopCap); cap.Raw() < gap.Raw() {
				gap = cap
			}
		}
		d.StopDist = gap
	}
	return d
}

// mean averages exactly, without going through float.
//
// The sum of a window of prices cannot overflow int64 at any size this accepts:
// a thousand bars of a million-dollar price is 1e17 raw, well inside the range.
func mean(xs []fixed.Dec) fixed.Dec {
	var total int64
	for _, x := range xs {
		total += x.Raw()
	}
	return fixed.FromRaw(total / int64(len(xs)))
}
